src/parser/jsonld.ts

```diff
@@ -0,0 +1,198 @@
+//
+// Copyright Inrupt Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal in
+// the Software without restriction, including without limitation the rights to use,
+// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
+// Software, and to permit persons to whom the Software is furnished to do so,
+// subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+//
+
+/* eslint-disable max-classes-per-file */
+import { fetch as defaultFetch } from "@inrupt/universal-fetch";
+import { promisifyEventEmitter } from "event-emitter-promisify";
+import type {
+  IJsonLdContext,
+  IParseOptions,
+  JsonLdContext,
+  JsonLdContextNormalized,
+} from "jsonld-context-parser";
+import { ContextParser, FetchDocumentLoader } from "jsonld-context-parser";
+import { JsonLdParser } from "jsonld-streaming-parser";
+import { Store } from "n3";
+import md5 from "md5";
+import type { JsonLd } from "../common/common";
+import CONTEXTS, { cachedContexts } from "./contexts";
+
+/**
+ * A JSON-LD document loader with the standard context for VCs pre-loaded
+ */
+export class CachedFetchDocumentLoader extends FetchDocumentLoader {
+  private contexts: Record<string, JsonLd>;
+
+  constructor(
+    contexts?: Record<string, JsonLd>,
+    private readonly allowContextFetching = false,
+    ...args: ConstructorParameters<typeof FetchDocumentLoader>
+  ) {
+    super(args[0] ?? defaultFetch);
+    this.contexts = { ...contexts, ...cachedContexts, ...CONTEXTS };
+  }
+
+  public async load(url: string): Promise<IJsonLdContext> {
+    if (Object.keys(this.contexts).includes(url)) {
+      return this.contexts[url as keyof typeof CONTEXTS];
+    }
+    if (!this.allowContextFetching) {
+      throw new Error(`Unexpected context requested [${url}]`);
+    }
+    return super.load(url);
+  }
+}
+
+export interface ParseOptions {
+  baseIRI?: string;
+  contexts?: Record<string, JsonLd>;
+  allowContextFetching?: boolean;
+}
+
+function hashOptions(options: IParseOptions | undefined) {
+  const opts = { ...options, parentContext: undefined };
+  for (const key of Object.keys(opts)) {
+    if (typeof opts[key as keyof typeof opts] === "undefined") {
```

Comment:
`Object.keys` always returns a `string[]` even if the set of keys defined in the type is more restricted. So in 73 there is a type error because `key` is a `string` but `opts` only accepts a finite set of keys.